### PR TITLE
Implementation of reorder and resize for touch devices.

### PR DESCRIPTION
**Pull request by SirGandal**
Enable reordering and resizing on touch devices. Based on the comments of @wcjordan on #236 this functionality should be turned on only if touchScrollEnabled is set to true.

## Description
- Addition of helper file _FixedDataTableEventHelper_ to extract x and y coordinates from either mouse or touch events.
- Extended _DOMMouseMoveTracker_ to account for touch events other than mouse events. Touch events trigger the same methods as their mouse equivalent,
- Added _touchEnabled_ property to _FixedDataTableCell_, _FixedDataTableCellGroup_, _FixedDataTableColumnResizeHandle_, _FixedDataTableColumnReorderHandle_ and _FixedDataTableRow_. None of these files are part of the public API so we should be fine. This is needed to propagate the _touchScrollEnabled_ property down to the _DOMMouseMoveTracker_ where we bind listeners to touch events.
- Added _touch-action: none;_ CSS rule to resize and reorder handles. This is needed since event.preventDefault() is used in a couple of places. Despite that works fine on mouse events, preventDefault logs a warning in console unless that rule is applied: "Unable to preventDefault inside passive event listener due to target being treated as passive."

## Motivation and Context
Since we are able to scroll on touch devices I believe that all the other functionalities (in this case reorder and resize) should work on touch devices as well.
Existing issue: #164 

## How Has This Been Tested?
Tested with device simulator in Chrome on Linux and Windows, on iPad running iOS 10.1.1 and Nexus 7 2013 running on Android 5.1.1.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.

### CODE REVIEW

File: examples/ReorderExample.js [hunk 1]
```diff
@@ -66,7 +66,7 @@ class ReorderExample extends React.Component {
     } else {
       if (fixedColumns.indexOf(event.reorderColumn) !== -1) {
         columnOrder.splice(fixedColumns.length - 1, 0, event.reorderColumn)
-      } else {
```
Comment (wcjordan) [inline examples/ReorderExample.js]: why was this change necessary?

File: src/FixedDataTableCell.js [hunk 2]
```diff
@@ -244,7 +249,15 @@ var FixedDataTableCell = createReactClass({
         <div
           className={cx('fixedDataTableCellLayout/columnResizerContainer')}
           style={columnResizerStyle}
-          onMouseDown={this._onColumnResizerMouseDown}>
+          /**
+           * Stop the event from being propagated when touching the handle.
+           * This prevents the rows from moving around when we resize the headers.
+           */
+          onTouchStart={e => e.stopPropagation()}
```
Comment (wcjordan) [inline src/FixedDataTableCell.js]: how does the div handle having 2 onTouchStart props?  Shouldn't we combine them?

File: src/FixedDataTableCell.js [hunk 3]
```diff
@@ -244,7 +249,15 @@ var FixedDataTableCell = createReactClass({
...
+          onTouchEnd={e => e.stopPropagation()}
+          onTouchMove={e => e.stopPropagation()}
```
Comment (wcjordan) [inline src/FixedDataTableCell.js]: use touchEnabled and only add these listeners if its true.

File: src/FixedDataTableColumnReorderHandle.js [hunk 4]
```diff
@@ -71,22 +77,32 @@ var FixedDataTableColumnReorderHandle = createReactClass({
           'fixedDataTableCellLayout/columnReorderContainer': true,
           'fixedDataTableCellLayout/columnReorderContainer/active': false,
         })}
+        /**
+         * Stop the event from being propagated when touching the handle.
+         * This prevents the rows from moving around when we drag the headers.
+         */
+        onTouchStart={e => e.stopPropagation()}
```
Comment (wcjordan) [inline src/FixedDataTableColumnReorderHandle.js]: Some questions about 2 onTouchStart handlers.

Also don't bind these events unless touchEnabled is true

File: src/css/layout/fixedDataTableCellGroupLayout.css [hunk 5]
```diff
@@ -27,4 +27,5 @@
 .fixedDataTableCellGroupLayout/cellGroupWrapper {
   position: absolute;
   top: 0;
+  touch-action: none;
```
Comment (wcjordan) [inline src/css/layout/fixedDataTableCellGroupLayout.css]: Since this could have unexpected consequences, I'm going to wait til v1.0 to pull it in.

If you want to rebase your changes off that branch, you should be able to use the beta release with it.  Otherwise, I'm open to taking the other changes, and you can add this css to your style sheet until v1 is released.

File: src/css/layout/fixedDataTableCellLayout.css [hunk 6]
```diff
@@ -50,6 +50,7 @@
   right: 0px;
   width: 6px;
   z-index: 1;
+  touch-action: none;
```
Comment (wcjordan) [inline src/css/layout/fixedDataTableCellLayout.css]: same as above

also what's the pro of putting this on individual elements rather than the whole table?

File: src/css/style/fixedDataTableCell.css [hunk 7]
```diff
@@ -23,6 +23,7 @@
 
 .public/fixedDataTableCell/cellContent {
   padding: 8px;
+  touch-action: none;
```
Comment (wcjordan) [inline src/css/style/fixedDataTableCell.css]: same as above

File: src/css/style/fixedDataTableColumnReorder.css [hunk 8]
```diff
@@ -8,6 +8,7 @@
   margin-right: -12px;
   float: left;
   cursor: move;
+  touch-action: none;
```
Comment (wcjordan) [inline src/css/style/fixedDataTableColumnReorder.css]: same as above

File: src/vendor_upstream/dom/DOMMouseMoveTracker.js [hunk 9]
```diff
@@ -52,34 +55,55 @@ class DOMMouseMoveTracker {
...
-        this._domNode,
-        'mouseleave',
-        this._onMouseEnd
-      );
-      this._eventOutToken = EventListener.listen(
-        this._domNode,
-        'mouseout',
-        this.onMouseEnd
-      );
+      if (!this._isTouchEnabled) {
```
Comment (wcjordan) [inline src/vendor_upstream/dom/DOMMouseMoveTracker.js]: I don't think this is an either or scenario, since some devices have both a touch screen and a mouse (e.g. a surface).

File: examples/ReorderExample.js (hunk 1 above)
Comment (SirGandal) [reply examples/ReorderExample.js]: My bad, this is leftover from some troubleshooting. Reverting.

File: src/FixedDataTableCell.js (hunk 2 above)
Comment (SirGandal) [reply src/FixedDataTableCell.js]: Not sure how I missed the 2 onTouchStart. I am going to combine them as suggested and stopPropagation only if touch is enabled since _onColumnResizerMouseDown is called onMouseDown as well.

File: src/FixedDataTableCell.js (hunk 3 above)
Comment (SirGandal) [reply src/FixedDataTableCell.js]: Good call.

File: src/FixedDataTableColumnReorderHandle.js (hunk 4 above)
Comment (SirGandal) [reply src/FixedDataTableColumnReorderHandle.js]: Same as above. I am going to combine them as suggested and stopPropagation only if touch is enabled since the onMouseDown function is called onMouseDown as well.

File: src/css/layout/fixedDataTableCellGroupLayout.css (hunk 5 above)
Comment (SirGandal) [reply src/css/layout/fixedDataTableCellGroupLayout.css]: Sounds good. Removing the _touch-action: none;_ rules from this PR. I will add that CSS to my stylesheet and wait for v1.

File: src/css/layout/fixedDataTableCellLayout.css (hunk 6 above)
Comment (SirGandal) [reply src/css/layout/fixedDataTableCellLayout.css]: That's a good point. I was trying to isolate the warning and apply the rule only to the elements that caused it.
Some quick testing by adding the rule to the entire table does show that it removes the warning completely and it does not impact other functionalities.

File: src/css/style/fixedDataTableCell.css (hunk 7 above)
Comment (SirGandal) [reply src/css/style/fixedDataTableCell.css]: Removing.

File: src/css/style/fixedDataTableColumnReorder.css (hunk 8 above)
Comment (SirGandal) [reply src/css/style/fixedDataTableColumnReorder.css]: Removing.

File: src/vendor_upstream/dom/DOMMouseMoveTracker.js (hunk 9 above)
Comment (SirGandal) [reply src/vendor_upstream/dom/DOMMouseMoveTracker.js]: That's fair, didn't think about that. Changing to always binding mouse events and only bind touch events when touch is enabled.

File: src/vendor_upstream/dom/DOMMouseMoveTracker.js [hunk 10]
```diff
@@ -72,14 +76,36 @@ class DOMMouseMoveTracker {
         'mouseout',
         this.onMouseEnd
       );
+
+      if (this._isTouchEnabled && !this._eventTouchStartToken &&
```
Comment (wcjordan) [inline src/vendor_upstream/dom/DOMMouseMoveTracker.js]: Should this be outside the if statement above?

File: examples/ReorderExample.js (hunk 1 above)
Comment (wcjordan) [reply examples/ReorderExample.js]: Don't forget this one

File: src/vendor_upstream/dom/DOMMouseMoveTracker.js (hunk 10 above)
Comment (SirGandal) [reply src/vendor_upstream/dom/DOMMouseMoveTracker.js]: Yes, good catch.

File: examples/ReorderExample.js (hunk 1 above)
Comment (SirGandal) [reply examples/ReorderExample.js]: Yup, forgot to push.

**Comment by SirGandal** [comment]
No problem, I appreciate the feedback 👍 
It looks good to me, you are free to merge, thanks!
